core/src/main/java/org/apache/calcite/rel/logical/LogicalProject.java

```diff
@@ -35,52 +37,63 @@
 import org.apache.calcite.util.Util;
 
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
 
 import org.checkerframework.checker.nullness.qual.Nullable;
 
 import java.util.List;
+import java.util.Objects;
+import java.util.Set;
 
 /**
  * Sub-class of {@link org.apache.calcite.rel.core.Project} not
  * targeted at any particular engine or calling convention.
  */
 public final class LogicalProject extends Project {
+  private final ImmutableSet<CorrelationId> variablesSet;
```

Comment:
I think it makes more sense to put `variablesSet` in the superclass (`Project`) rather than here.

Reply:
Does `variablesSet` belong to `Project` at all? I thought it was `Correlate` that sets variables.

Reply:
@vlsi  I thought so too but apparently `variablesSet` is also on `LogicalFilter`, `Join`, and `Correlate`.

Reply:
Oh, it looks like https://github.com/apache/calcite/commit/505a9064b96a6c8399735fc2fa4d0ac9d5f3ed87 and https://issues.apache.org/jira/browse/CALCITE-1045 are relevant.

Reply:
It does make sense for project with correlated scalar queries. For example, the query like `SELECT t0.v, (SELECT t0.v + t1.v FROM test_tbl t1) AS j FROM test_tbl t0` has a logical plan:
```
LogicalProject(V=[$2], J=[$SCALAR_QUERY({
LogicalProject(T0.V + T1.V=[+($cor0.V, $2)])
  IgniteLogicalTableScan(table=[[PUBLIC, TEST_TBL]])
})])
  IgniteLogicalTableScan(table=[[PUBLIC, TEST_TBL]])
```
In order to evaluate this plan you need either implement filter that set correlation variable, or convert it to LogicalCorrelate (what is actually done by `SubQueryRemoveRule` ).

With that said, it seems to me LogicalFilter has to expose the variables they set, but not every physical implementation would do so.

Reply:
@korlov42 , thanks for the clarifications. I see. I thought we did not have that in Calcite (I thought Correlate was the only way to set correlate vars).

Reply:
So, I moved `variableSet` to a `Project` class in order to explain this term before the regular project fields